### PR TITLE
Feature/sqlalchemy dataset

### PR DESCRIPTION
**Pull request by jpcampb2**
This PR creates an initial SqlAlchemyDataSet with the following implemented expectations. 
See #12 
 - `expect_column_to_exist`
 - `expect_table_row_count_to_be`
 - `expect_table_row_count_to_be_between`
 - `expect_column_values_to_not_be_null`
 - `expect_column_values_to_be_null`
 - `expect_column_values_to_be_in_set`
 - `expect_column_values_to_be_between`
 - `expect_column_mean_to_be`
 - `expect_column_min_to_be`
 - `expect_column_max_to_be`
 - `expect_column_sum_to_be`

The PR also introduces the new system of testing expectations for many expectations to ensure as much parity across systems as possible. See #167

### CODE REVIEW

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+
+
+class MetaSqlAlchemyDataSet(DataSet):
+
+    def __init__(self, *args, **kwargs):
+        super(MetaSqlAlchemyDataSet, self).__init__(*args, **kwargs)
+
+    @classmethod
+    def column_map_expectation(cls, func):
+        """For SqlAlchemy, this decorator allows individual column_map_expectations to simply return the filter
```
Comment (abegong) [inline great_expectations/dataset/sqlalchemy_dataset.py:21]: 💯

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+                unexpected_count_limit = result_format['partial_unexpected_count']
+            else:
+                if result_format['result_obj_format'] == 'COMPLETE':
+                    unexpected_count_limit = None
+                else:
+                    unexpected_count_limit = 20
+
+            unexpected_condition = func(self, column, *args, **kwargs)
+
+            count_query = select([
```
Comment (abegong) [inline great_expectations/dataset/sqlalchemy_dataset.py]: Nice consolidation!

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+                    case([(sa_column(column) == None, 1)], else_=0)
+                ).label('null_count'),
+                sa_func.sum(
+                    case([(unexpected_condition, 1)], else_=0)
+                ).label('unexpected_count')
+            ]).select_from(table(self.table_name))
+
+            count_results = self.engine.execute(count_query).fetchone()
+
+            unexpected_query_results = self.engine.execute(
```
Comment (abegong) [inline great_expectations/dataset/sqlalchemy_dataset.py]: This is very cool. I hadn't realized that we would be able to be so direct.

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+                else:
+                    unexpected_count_limit = 20
+
+
+            evaluation_result = func(self, column, *args, **kwargs)
+
+            if ('success' not in evaluation_result) or \
+                ('result_obj' not in evaluation_result) or \
+                ('observed_value' not in evaluation_result['result_obj']):
+                raise ValueError("Column aggregate expectation failed to return required return information.")
```
Comment (abegong) [inline great_expectations/dataset/sqlalchemy_dataset.py]: Minor nit: this error message could be more informative...

Comment (abegong) [reply great_expectations/dataset/sqlalchemy_dataset.py]: (Fixing this probably doesn't belong on this PR.)

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+        """
+        for col in self.columns:
+            self.append_expectation({
+                "expectation_type": "expect_column_to_exist",
+                "kwargs": {
+                    "column": col["name"]
+                }
+            })
+
+    def _is_numeric_column(self, column):
```
Comment (abegong) [inline great_expectations/dataset/sqlalchemy_dataset.py:174]: ?

Comment (abegong) [reply great_expectations/dataset/sqlalchemy_dataset.py:174]: hm. We've brought typing in via the backdoor.

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+    ):
+
+        if min_value is None and max_value is None:
+            raise ValueError("min_value and max_value cannot both be None")
+
+        if parse_strings_as_datetimes:
+            raise ValueError("parse_strings_as_datetimes is not supported in SqlAlchemy")
+
+        col_max = self.engine.execute(
+            select([sa_func.max(sa_column(column))]).select_from(table(self.table_name))
```
Comment (abegong) [inline great_expectations/dataset/sqlalchemy_dataset.py]: Got it. So `column_aggregate_expectations` act almost exactly like they do in `PandasDataSet`.

Side thought: we need to update our docs for custom expectations.

File: great_expectations/dataset/util.py [hunk 7]
```diff
@@ -1,18 +1,25 @@
...
 import warnings
 import sys
 import copy
-import json
 import datetime
 
 from functools import wraps
 
 
+def parse_result_format(result_format):
```
Comment (abegong) [inline great_expectations/dataset/util.py:18]: ? A docstring would be helpful here.

File: tests/column_aggregate_expectations/test_column_aggregate_expectations.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
+###
+###
+#
+# This file should not be modified. To adjust test cases, edit the related json file.
```
Comment (abegong) [inline tests/column_aggregate_expectations/test_column_aggregate_expectations.py]: "file(s)"

File: tests/test_pandas_dataset.py [hunk 9]
```diff
@@ -10,108 +10,26 @@
...
 class TestPandasDataset(unittest.TestCase):
 
-    def test_expect_column_to_exist(self):
-        print("=== test_expect_column_to_exist ===")
-        with open("./tests/test_sets/expect_column_to_exist_test_set.json") as f:
-            J = json.load(f)
-            D = ge.dataset.PandasDataSet(J["dataset"])
-            D.set_default_expectation_argument("result_format", "COMPLETE")
-            T = J["tests"]
+    # def test_expect_column_to_exist(self):
```
Comment (abegong) [inline tests/test_pandas_dataset.py]: We might as well decruft all these commented tests, don't you think?

**Review by abegong — CHANGES_REQUESTED** [review]
This is awesome work. One of my favorite PRs so far.

I have a few small changes to request, but on the whole, this set of changes feels really, really good.

File: great_expectations/dataset/sqlalchemy_dataset.py (hunk 5 above)
Comment (jpcampb2) [reply great_expectations/dataset/sqlalchemy_dataset.py:174]: True. I'd argue it's no more severe than what happens right now with the implied type coercions in python (that we inherit in PandasDataSet). Actually, this was necessary to ensure we're staying closer between the two right now. But I'm definitely looking forward to a more elegant solution, e.g. as implied in #207 .

File: great_expectations/dataset/sqlalchemy_dataset.py (hunk 6 above)
Comment (jpcampb2) [reply great_expectations/dataset/sqlalchemy_dataset.py]: Good catch. I updated docs for custom expectations and also added an example of a custom SqlAlchemyDataSet that uses both kinds of decorators.

File: tests/column_aggregate_expectations/test_column_aggregate_expectations.py (hunk 8 above)
Comment (jpcampb2) [reply tests/column_aggregate_expectations/test_column_aggregate_expectations.py]: 👍 :)

File: tests/test_pandas_dataset.py (hunk 9 above)
Comment (jpcampb2) [reply tests/test_pandas_dataset.py]: Agreed. Pulled them out.

File: great_expectations/dataset/sqlalchemy_dataset.py (hunk 5 above)
Comment (abegong) [reply great_expectations/dataset/sqlalchemy_dataset.py:174]: :+1:

File: great_expectations/dataset/util.py (hunk 7 above)
Comment (jpcampb2) [reply great_expectations/dataset/util.py:18]: Done.

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 10]
```diff
@@ -297,7 +293,7 @@ def expect_column_values_to_be_null(self,
         result_format=None, include_config=False, catch_exceptions=None, meta=None
     ):
 
-        return sa_column(column) != None
+        return sa.column(column) == None
```
Comment (abegong) [inline great_expectations/dataset/sqlalchemy_dataset.py]: Yikes. Good catch.

**Review by abegong — APPROVED** [review]
This looks great!

**Comment by jpcampb2** [comment]
Woohoo!

**Comment by abegong** [comment]
:tada::tada::tada:!

**Pull request merged**